packages/nextjs/utils/scaffold-eth/common.ts

```diff
@@ -1,3 +1,29 @@
+import type { MergeDeepRecord } from "type-fest/source/merge-deep";
+
 // To be used in JSON.stringify when a field might be bigint
 // https://wagmi.sh/react/faq#bigint-serialization
-export const replacer = (key: string, value: unknown) => (typeof value === "bigint" ? value.toString() : value);
+export const replacer = (_key: string, value: unknown) => (typeof value === "bigint" ? value.toString() : value);
+
+export const deepMerge = <D extends Record<PropertyKey, any>, S extends Record<PropertyKey, any>>(
+  destination: D,
+  source: S,
+) => {
+  const result: Record<PropertyKey, any> = {};
+  for (const key in destination) {
+    if (key in source) {
+      if (typeof destination[key] === "object" && typeof source[key] === "object" && !Array.isArray(source[key])) {
```

Comment:
Summary how this function works is it recursively merges the value of the key if its an `object` and will directly override `destination` key if its value is primitive type by the value of the same key in the `source`

The reason I have `!Array.isArray(source[key])` here is because there might be an edge case where suppose : 

`deployedContrac.ts`: 

```ts
{
  1: {
    YourContract : {address : "0xabcd", abi : [YourContract_ABI] }
  }
}
```

`externalContract.ts`: 

```ts
// Bad guy trying to break SE-2 
{
  1: {
    YourContract : {address : "daiAddress", abi : [DAI_ABI] }
  }
}
```

So the final result will be, Instead of recursively merging Abi's we just replace with `externalContract` abi : 

```ts
{
  1: {
    YourContract : {address : "daiAddress", abi : [DAI_ABI] }
  }
}
```

 If we don't do that it will break ABI's array : 

```ts
{
  1: {
    YourContract : {address : "daiAddress", abi : [DISASTER] }
  }
}
```

The reason for `DISASTER` is because since `typeof Array` is `object` this function will try to recursively merge each element of both ABI's 

Hence if its array type we just override `destination` key with `source` key's array value similar to what we do with other primitive types

Reply:
If I understand right, you're merging objects only on first two levels - networkId and contractNames. So, can we use something simpler? Not sure about types though. Something like that
```ts
  const result: Record<PropertyKey, any> = {};
  const allKeys = Array.from(new Set([...Object.keys(source), ...Object.keys(destination)]));
  for (const key of allKeys) {
    result[key] = { ...destination[key], ...source[key] };
  }
  return result as MergeDeepRecord<D, S, { arrayMergeMode: "replace" }>;
```

Reply:
I actually wrote generic logic in line with how `MergeDeepRecord` works but I think it not needed,  and  I really love the proposed solution minimal, inline with `MergeDeepRecord`   and handles our required edge-cases too !

Updated at https://github.com/scaffold-eth/scaffold-eth-2/pull/592/commits/f47535cf0a007ac39df864a98bb4217da62c829a  tysm Rinat 🙌